### DISCUSSION TITLE
Setting storage for Dialogs (forms)

### DISCUSSION THREAD

**Discussion by reach2rv** [serenity-is/Serenity#5872] [Serenity Ideas]
Hello @VictorTomaili @minhhungit @dfaruque 

How about having setting storage for dialogs as well, same like grid persistence storage. With this user (admin) can define which fields to show to end user ( we can also set default values) . by doing this application can become configurable incase of one application to multiple customers and customization can be avoided.

Please put forward your suggestions.

**Comment by minhhungit** [comment]
IMHO I don't think that is healthy, I normally return all fields by form.cs and write code in afterLoadEntity to decide to hiding or showing some fields

If there is many difference about logic between "versions" of the form, consider create more form/dialog for this case

**Reply by reach2rv** [reply]
I am following the same. but there are scenarios where one form has few fields which are not mandatory( specific to customer or region) thus need to be shown or hide as per the settings.

I am currently planning to use following approach

```TypeScript
getPropertyItems() {
            var PropertyItems = super.getPropertyItems();
            var editor = PropertyItems.filter(x => x.name == 'PanNumber')[0];
            editor.Hidden= true;           
            return PropertyItems;
        }
```

can you guide me how I can set values of persistence setting to interface?  so I can map that with form and apply setting

**Reply by reach2rv** [reply]
```TypeScript
getPropertyItems() {
            var PropertyItems = super.getPropertyItems();
            var Setting = this.getPersistedDialogSettings();
            Setting.Fields.forEach(function (field) {
                var editor = PropertyItems.filter(x => x.name == field.id)[0];
                if (field.value != null)
                    editor.defaultValue = 'ABC';
                editor.visible = field.visible;
                editor.readOnly = field.readonly;
                editor.required = field.required;
            });           
            return PropertyItems;
        }


        getPersistedDialogSettings(): Serenity.PersistedDialogSettings {
            var storage = new TabAurous.Common.UserPreferenceStorage();
            if (storage == null)
                return null;

            var json = Q.trimToNull(storage.getItem('POSForm'));
            if (json != null && Q.startsWith(json, '{') && Q.endsWith(json, '}'))
                return JSON.parse(json);

            return null;
        }
```

I will add new form to set this settings in storage with Key (mostly will use form key to avoid conflict).

**Reply by edwardch** [reply]
If that's your scenario, how about adding custom read permission to those fields so you can manage it on a user or role level?

I've done this so some users get to see certain fields, others do not.

**Reply by reach2rv** [reply]
@edwardch this can be implemented with one app per customer scenario. My scenario is one app for multiple customers. though they all fall under same business vertical, still requires some changes to suffice their specific needs. implementing this will help me to handle this with single code base ( we do have production and development branch ) without creating branch or different code base for each customer.

I have almost completed the client side implementation will be working on setting up and saving form data next week.. will keep this thread updated so if anyone needs such kind of implementation can utilize this
